stl/src/locale0.cpp

```diff
@@ -135,8 +135,39 @@ __PURE_APPDOMAIN_GLOBAL locale::id ctype<unsigned short>::id(0);
 
 __PURE_APPDOMAIN_GLOBAL locale::id codecvt<unsigned short, char, mbstate_t>::id(0);
 
+#define _Compiler_barrier() _STL_DISABLE_DEPRECATED_WARNING _ReadWriteBarrier() _STL_RESTORE_DEPRECATED_WARNING
+
+#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_ARM64EC)
+#define _Memory_barrier()             __dmb(0xB) // inner shared data memory barrier
+#define _Compiler_or_memory_barrier() _Memory_barrier()
+#elif defined(_M_IX86) || defined(_M_X64)
+// x86/x64 hardware only emits memory barriers inside _Interlocked intrinsics
+#define _Compiler_or_memory_barrier() _Compiler_barrier()
+#else // ^^^ x86/x64 / unsupported hardware vvv
+#error Unsupported hardware
+#endif // hardware

```

Comment:
we might be able to get away with moving this from atomic into xatomic and including that here. The only possible codegen there is the inline constexpr variables and some templates that are very, very simple. Neither should normally generate code.

Reply:
Since I don't have access to the `/clr:pure` tests, would it be too much to ask for someone to try just `include <xatomic.h>` in locale0.cpp as a smoke test?

Reply:
So, it turns out that `<atomic>` is transitively included via `<locale>` and `#undef`s the macros we want:
https://github.com/microsoft/STL/blob/f9697fc439f863c444c729ad6ffbddd574797cda/stl/inc/atomic#L2981-L2983
So we either have to unconditionally not `#undef` them, and risk releasing them into the wild, or add an `#ifdef` section that will selectively not `#undef` them just for locale0.cpp. Any preference?

Reply:
We recently had a ton of trouble with `<atomic>` being dragged into `filesystem.cpp` and therefore the import lib - an experience I am not eager to repeat :joy_cat: I think I would recommend simply duplicating this code rather than attempting to centralize it. (That is, `locale0.cpp`'s "keep this as small as possible" warning comment is still very important.)

Thus, unless I am missing something, I think that we should take no action here. (In principle I do agree with Charlie's observation that centralization would be preferable, I just think that attempting to do so would cause too many headaches.)

Reply:
@StephanTLavavej It sounds like you're advising against doing Charlie's only suggestions, so I don't think there are any outstanding comments for me to resolve.

Reply:
Yep, we talked about this at the weekly maintainer meeting and have filed #3103 as a followup issue - nothing to do for this PR right now. Thanks!